qp_deblur/deblur.py

```diff
@@ -157,6 +158,145 @@ def generate_sepp_placements(seqs, out_dir, threads, reference_phylogeny=None,
+        placements = json.loads(f.read())
+        placements['placements'] = []
+        with open(file_template, 'w') as fw:
+            json.dump(placements, fw)
+
+    return (file_template, '%s/dummy_rename-json.py' % out_dir)
+
+
+def generate_insertion_trees(placements, out_dir,
+                             reference_template=None,
+                             reference_rename=None):
+    """Generates phylogenetic trees by inserting placements into a reference
+
+    Parameters
+    ----------
+    placements : dict of strings
+        keys are the seqs, values are the new placements as JSON strings
+    out_dir : str
+        The job output directory
+    reference_template : str, optional
+        Filepath to the reference placement json file.
+        This file can be produced via _generate_template_rename() and should be
+        stored in the plugin package, because it can re used.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+    reference_rename : str, optional
+        Similar to reference_template, but a filepath to the generated python
+        renaming script to undo the name scaping post guppy.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+
+    Returns
+    -------
+    str
+        The filepath of the phylogenetic insertion tree in Newick format.
+    """
+
+    # create a valid placement.json file as input for guppy
+    file_ref_template = resource_filename(
+        Requirement.parse('qp_deblur'),
+        'qp_deblur/assets/tmpl_gg13.8-99_placement.json')
+    if reference_template is not None:
+        file_ref_template = reference_template
+    if not exists(file_ref_template):
+        raise ValueError("Reference template '%s' does not exits!" %
+                         file_ref_template)
+    with open(file_ref_template, 'r') as f:
+        plcmnts = json.loads(f.read())
+        for sequence, placement in placements.items():
+            plcmnts['placements'].append({'p': placement,
+                                          'nm': [[sequence, 1]]})
+    file_placements = '%s/placements.json' % out_dir
+    with open(file_placements, 'w') as f:
+        json.dump(plcmnts, f)
+
+    # execute guppy
+    file_tree_escaped = join(out_dir, 'insertion_tree.tre')
+    std_out, std_err, return_value = system_call(
+        'guppy tog %s -o %s' % (file_placements, file_tree_escaped))
+    if return_value != 0:
+        error_msg = ("Error running guppy:\nStd out: %s\nStd err: %s"
+                     % (std_out, std_err))
+        raise ValueError(error_msg)
+
+    # execute node name re-labeling (to revert the escaping of names necessary
+    # for guppy)
+    file_ref_rename = resource_filename(
+        Requirement.parse('qp_deblur'),
+        'qp_deblur/assets/tmpl_gg13.8-99_rename-json.py')
+    if reference_rename is not None:
+        file_ref_rename = reference_rename
+    if not exists(file_ref_rename):
```

Comment:
could you perform these checks at the beginning of the function? This way we don't run guppy at all if this fails, otherwise is wasted compute.

Reply:
good point! Changed